File: backend/routes/tasks.py
from flask import Blueprint, request, jsonify, g
import uuid
from datetime import datetime
from db import get_db
from auth import auth_required

tasks_bp = Blueprint('tasks', __name__)
tasks_bp.before_request(auth_required)

def gen_id():
    d = datetime.now()
    return f"t_{d.strftime('%Y%m%d')}_{uuid.uuid4().hex[:8]}"

def row_to_task(row):
    return {
        'task_id': row['task_id'],
        'title': row['title'],
        'notes': row['notes'] or '',
        'due_date': row['due_date'],
        'created_at': row['created_at'],
        'repeat': row['repeat_type'],
        'repeat_end': row['repeat_end'],
        'status': row['status'],
        'snooze_count': row['snooze_count'],
        'alarm_enabled': bool(row['alarm_enabled']),
        'completed_at': row['completed_at'],
        'instance_completions': __import__('json').loads(row['instance_completions'] or '{}'),
        'category': row['category'],
    }

@tasks_bp.route('/', methods=['GET'])
def get_tasks():
    db = get_db()
    rows = db.execute('SELECT * FROM tasks WHERE user_id = ? ORDER BY due_date ASC', (g.user_id,)).fetchall()
    db.close()
    return jsonify({'tasks': [row_to_task(r) for r in rows]})

@tasks_bp.route('/', methods=['POST'])
def create_task():
    data = request.get_json()
    title = (data.get('title') or '').strip()
    due_date = data.get('due_date')

    if not title:
        return jsonify({'error': 'Task title is required.'}), 400
    if not due_date:
        return jsonify({'error': 'Due date is required.'}), 400

    task_id = gen_id()
    now = datetime.utcnow().isoformat()
    db = get_db()
    db.execute(
        'INSERT INTO tasks (user_id, task_id, title, notes, due_date, created_at, repeat_type, category) VALUES (?, ?, ?, ?, ?, ?, ?, ?)',
        (g.user_id, task_id, title, data.get('notes', ''), due_date, now, data.get('repeat', 'none'), data.get('category', 'General'))
    )
    db.commit()
    row = db.execute('SELECT * FROM tasks WHERE task_id = ?', (task_id,)).fetchone()
    db.close()
    return jsonify({'task': row_to_task(row)}), 201

@tasks_bp.route('/<task_id>', methods=['PUT'])
def update_task(task_id):
    db = get_db()
    row = db.execute('SELECT * FROM tasks WHERE task_id = ? AND user_id = ?', (task_id, g.user_id)).fetchone()
    if not row:
        db.close()
        return jsonify({'error': 'Task not found.'}), 404

    data = request.get_json()
    fields = []
    values = []

    if 'title' in data and data['title'] is not None:
        fields.append('title = ?')
        values.append(data['title'].strip())
    if 'notes' in data and data['notes'] is not None:
        fields.append('notes = ?')
        values.append(data['notes'])
    if 'due_date' in data and data['due_date'] is not None:
        fields.append('due_date = ?')
        values.append(data['due_date'])
    if 'repeat' in data and data['repeat'] is not None:
        fields.append('repeat_type = ?')
        values.append(data['repeat'])
    if 'category' in data and data['category'] is not None:
        fields.append('category = ?')
        values.append(data['category'])
    if 'status' in data and data['status'] is not None:
        fields.append('status = ?')
        values.append(data['status'])
        if data['status'] == 'completed':
            fields.append('completed_at = ?')
            values.append(datetime.utcnow().isoformat())
        if data['status'] == 'pending':
            fields.append('completed_at = ?')
            values.append(None)
    if 'alarm_enabled' in data and data['alarm_enabled'] is not None:
        fields.append('alarm_enabled = ?')
        values.append(1 if data['alarm_enabled'] else 0)

    if not fields:
        db.close()
        return jsonify({'error': 'No fields to update.'}), 400

    values.extend([task_id, g.user_id])
    db.execute(f"UPDATE tasks SET {', '.join(fields)} WHERE task_id = ? AND user_id = ?", values)
    db.commit()
    updated = db.execute('SELECT * FROM tasks WHERE task_id = ?', (task_id,)).fetchone()
    db.close()
    return jsonify({'task': row_to_task(updated)})

@tasks_bp.route('/<task_id>/toggle', methods=['POST'])
def toggle_task(task_id):
    import json
    db = get_db()
    row = db.execute('SELECT * FROM tasks WHERE task_id = ? AND user_id = ?', (task_id, g.user_id)).fetchone()
    if not row:
        db.close()
        return jsonify({'error': 'Task not found.'}), 404

    if row['status'] == 'completed':
        db.execute("UPDATE tasks SET status = 'pending', completed_at = NULL WHERE task_id = ?", (task_id,))
    elif row['repeat_type'] != 'none':
        data = request.get_json() or {}
        completions = json.loads(row['instance_completions'] or '{}')
        date_key = data.get('dateKey', datetime.utcnow().strftime('%Y-%m-%d'))
        completions[date_key] = 'completed'
        db.execute('UPDATE tasks SET instance_completions = ? WHERE task_id = ?', (json.dumps(completions), task_id))
    else:
        db.execute("UPDATE tasks SET status = 'completed', completed_at = ? WHERE task_id = ?", (datetime.utcnow().isoformat(), task_id))

    db.commit()
    updated = db.execute('SELECT * FROM tasks WHERE task_id = ?', (task_id,)).fetchone()
    db.close()
    return jsonify({'task': row_to_task(updated)})

@tasks_bp.route('/<task_id>/snooze', methods=['POST'])
def snooze_task(task_id):
    data = request.get_json()
    minutes = data.get('minutes', 0)
    if not minutes or minutes <= 0:
        return jsonify({'error': 'Minutes must be > 0.'}), 400

    db = get_db()
    row = db.execute('SELECT * FROM tasks WHERE task_id = ? AND user_id = ?', (task_id, g.user_id)).fetchone()
    if not row:
        db.close()
        return jsonify({'error': 'Task not found.'}), 404

    due = datetime.fromisoformat(row['due_date'].replace(' ', 'T'))
    from datetime import timedelta
    new_due = due + timedelta(minutes=minutes)
    new_due_str = new_due.strftime('%Y-%m-%d %H:%M:%S')

    db.execute('UPDATE tasks SET snooze_count = snooze_count + 1, due_date = ? WHERE task_id = ?', (new_due_str, task_id))
    db.commit()
    updated = db.execute('SELECT * FROM tasks WHERE task_id = ?', (task_id,)).fetchone()
    db.close()
    return jsonify({'task': row_to_task(updated)})

@tasks_bp.route('/<task_id>', methods=['DELETE'])
def delete_task(task_id):
    db = get_db()
    result = db.execute('DELETE FROM tasks WHERE task_id = ? AND user_id = ?', (task_id, g.user_id))
    db.commit()
    changes = result.rowcount
    db.close()
    if changes == 0:
        return jsonify({'error': 'Task not found.'}), 404
    return jsonify({'message': 'Deleted.'})

@tasks_bp.route('/', methods=['DELETE'])
def delete_all_tasks():
    db = get_db()
    result = db.execute('DELETE FROM tasks WHERE user_id = ?', (g.user_id,))
    db.commit()
    changes = result.rowcount
    db.close()
    return jsonify({'message': f'Deleted {changes} task(s).'})

File: backend/routes/tasks.js
const express = require('express');
const crypto = require('crypto');
const db = require('../db');
const { auth } = require('../middleware');

const router = express.Router();

router.use(auth);

function genId() {
  const now = new Date();
  const datePart = `${now.getFullYear()}${String(now.getMonth() + 1).padStart(2, '0')}${String(now.getDate()).padStart(2, '0')}`;
  return `t_${datePart}_${crypto.randomBytes(4).toString('hex')}`;
}

function rowToTask(row) {
  return {
    task_id: row.task_id,
    title: row.title,
    notes: row.notes || '',
    due_date: row.due_date,
    created_at: row.created_at,
    repeat: row.repeat_type,
    repeat_end: row.repeat_end,
    status: row.status,
    snooze_count: row.snooze_count,
    alarm_enabled: !!row.alarm_enabled,
    completed_at: row.completed_at,
    instance_completions: JSON.parse(row.instance_completions || '{}'),
    category: row.category,
  };
}

// GET /tasks
router.get('/', (req, res) => {
  const rows = db.prepare('SELECT * FROM tasks WHERE user_id = ? ORDER BY due_date ASC').all(req.userId);
  res.json({ tasks: rows.map(rowToTask) });
});

// POST /tasks
router.post('/', (req, res) => {
  const { title, notes, due_date, repeat, category } = req.body;
  if (!title || !title.trim()) {
    return res.status(400).json({ error: 'Task title is required.' });
  }
  if (!due_date) {
    return res.status(400).json({ error: 'Due date is required.' });
  }

  const taskId = genId();
  const now = new Date().toISOString();
  db.prepare(`
    INSERT INTO tasks (user_id, task_id, title, notes, due_date, created_at, repeat_type, category)
    VALUES (?, ?, ?, ?, ?, ?, ?, ?)
  `).run(req.userId, taskId, title.trim(), notes || '', due_date, now, repeat || 'none', category || 'General');

  const row = db.prepare('SELECT * FROM tasks WHERE task_id = ?').get(taskId);
  res.status(201).json({ task: rowToTask(row) });
});

// PUT /tasks/:taskId
router.put('/:taskId', (req, res) => {
  const row = db.prepare('SELECT * FROM tasks WHERE task_id = ? AND user_id = ?').get(req.params.taskId, req.userId);
  if (!row) return res.status(404).json({ error: 'Task not found.' });

  const { title, notes, due_date, repeat, category, status, alarm_enabled } = req.body;

  const fields = [];
  const values = [];

  if (title !== undefined) { fields.push('title = ?'); values.push(title.trim()); }
  if (notes !== undefined) { fields.push('notes = ?'); values.push(notes); }
  if (due_date !== undefined) { fields.push('due_date = ?'); values.push(due_date); }
  if (repeat !== undefined) { fields.push('repeat_type = ?'); values.push(repeat); }
  if (category !== undefined) { fields.push('category = ?'); values.push(category); }
  if (status !== undefined) {
    fields.push('status = ?');
    values.push(status);
    if (status === 'completed') { fields.push('completed_at = ?'); values.push(new Date().toISOString()); }
    if (status === 'pending') { fields.push('completed_at = ?'); values.push(null); }
  }
  if (alarm_enabled !== undefined) { fields.push('alarm_enabled = ?'); values.push(alarm_enabled ? 1 : 0); }

  if (fields.length === 0) {
    return res.status(400).json({ error: 'No fields to update.' });
  }

  values.push(req.params.taskId, req.userId);
  db.prepare(`UPDATE tasks SET ${fields.join(', ')} WHERE task_id = ? AND user_id = ?`).run(...values);

  const updated = db.prepare('SELECT * FROM tasks WHERE task_id = ?').get(req.params.taskId);
  res.json({ task: rowToTask(updated) });
});

// POST /tasks/:taskId/toggle
router.post('/:taskId/toggle', (req, res) => {
  const row = db.prepare('SELECT * FROM tasks WHERE task_id = ? AND user_id = ?').get(req.params.taskId, req.userId);
  if (!row) return res.status(404).json({ error: 'Task not found.' });

  const task = rowToTask(row);

  if (task.status === 'completed') {
    db.prepare("UPDATE tasks SET status = 'pending', completed_at = NULL WHERE task_id = ?").run(row.task_id);
  } else if (task.repeat !== 'none') {
    const { dateKey } = req.body;
    const completions = JSON.parse(row.instance_completions || '{}');
    completions[dateKey || new Date().toISOString().slice(0, 10)] = 'completed';
    db.prepare('UPDATE tasks SET instance_completions = ? WHERE task_id = ?').run(JSON.stringify(completions), row.task_id);
  } else {
    db.prepare("UPDATE tasks SET status = 'completed', completed_at = ? WHERE task_id = ?").run(new Date().toISOString(), row.task_id);
  }

  const updated = db.prepare('SELECT * FROM tasks WHERE task_id = ?').get(row.task_id);
  res.json({ task: rowToTask(updated) });
});

// POST /tasks/:taskId/snooze
router.post('/:taskId/snooze', (req, res) => {
  const { minutes } = req.body;
  if (!minutes || minutes <= 0) return res.status(400).json({ error: 'Minutes must be > 0.' });

  const row = db.prepare('SELECT * FROM tasks WHERE task_id = ? AND user_id = ?').get(req.params.taskId, req.userId);
  if (!row) return res.status(404).json({ error: 'Task not found.' });

  const due = new Date(row.due_date);
  due.setMinutes(due.getMinutes() + minutes);
  const newDue = due.toISOString().replace('T', ' ').slice(0, 19);

  db.prepare('UPDATE tasks SET snooze_count = snooze_count + 1, due_date = ? WHERE task_id = ?').run(newDue, row.task_id);

  const updated = db.prepare('SELECT * FROM tasks WHERE task_id = ?').get(row.task_id);
  res.json({ task: rowToTask(updated) });
});

// DELETE /tasks/:taskId
router.delete('/:taskId', (req, res) => {
  const result = db.prepare('DELETE FROM tasks WHERE task_id = ? AND user_id = ?').run(req.params.taskId, req.userId);
  if (result.changes === 0) return res.status(404).json({ error: 'Task not found.' });
  res.json({ message: 'Deleted.' });
});

// DELETE /tasks  (delete all for user)
router.delete('/', (req, res) => {
  const result = db.prepare('DELETE FROM tasks WHERE user_id = ?').run(req.userId);
  res.json({ message: `Deleted ${result.changes} task(s).` });
});

module.exports = router;
